tests/framework/ab_test.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Defines utilities for performing A/B-tests.
+
+A/B-Tests are style of tests where we do not care what state a test is in, but only that this state does not change
+across a pull request. This is useful if
+1. Validating the state requires some baseline to be persisted in the repository, and maintaining this baseline
+   adds significant operational burden (for example, performance tests), or
+2. The state can change due to outside factors (e.g. Hardware changes), and such external changes would block all
+   pull requests until they are resolved.
+
+Consider for example a `cargo audit` tests, which is used to reject usage of dependency versinos that have known
+security vulnerabilities, or which have been yanked. The "state" here is "list of vulnerable dependencies". Clearly,
+this can change due to external action (a new vulnerability is discovered and published to RustSec). At this point,
+every PR would fail until this dependency is removed, blocking all development. Simply removing the test from PR CI
+is not an option, since we want to avoid the scenario where a PR adds a dependency with a known vulnerability (e.g.
+the PR itself changes the "list of vulnerable dependencies"). A/B-Testing allows us to not block PRs on the former case,
+while still preventing the latter: We run cargo audit twice, once on main HEAD, and once on the PR HEAD. If the output
+of both invocations is the same, the test passes (with us being alerted to this situtation via a special pipeline that
+does not block PRs). If not, it fails, preventing PRs from introducing new vulnerable dependencies.
+"""
+import contextlib
+import os
+from tempfile import TemporaryDirectory
+from typing import Callable, Optional, TypeVar
+
+from framework import utils
+
+# Locally, this will always compare against main, even if we try to merge into, say, a feature branch.
+# We might want to do a more sophisticated way to determine a "parent" branch here.
+DEFAULT_A_REVISION = os.environ.get("BUILDKITE_PULL_REQUEST_BASE_BRANCH", "main")
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def default_comparator(ah: T, be: T) -> bool:
+    """Returns `true` iff that the two arguments are equal.
+
+    The default assertion for A/B-tests using `ab_test`.
+
+    Ridiculous variable names sponsored by pylint."""
+    return ah == be
```

Comment:
nit:

> Ridiculous variable names sponsored by pylint.

This comment made me laugh :þ

How about `a_out` and `b_out`, since outputs assume to be given to them?
Or `a_result` and `b_result` like those in `git_ab_test()`?